src/protocol/codecs/algomsg.rs

```diff
@@ -0,0 +1,155 @@
+use std::io;
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+use tracing::*;
+
+use super::topic::TopicCodec;
+use crate::protocol::invalid_data;
+
+/// [Tag] represents a message type identifier.
+///
+/// The original tag list can be found in go-algorand/protocol/tags.go.
+#[derive(Debug)]
+pub enum Tag {
+    UnknownMsg,
+    AgreementVote,
+    MsgOfInterest,
+    MsgDigestSkip,
+    NetPrioResponse,
+    Ping,
+    PingReply,
+    ProposalPayload,
+    StateProofSig,
+    TopicMsgResp,
+    Txn,
+    UniCatchupReq,
+    UniEnsBlockReq,
+    VoteBundle,
+}
+
+impl TryFrom<BytesMut> for Tag {
+    type Error = io::Error;
+
+    fn try_from(tag: BytesMut) -> Result<Self, Self::Error> {
+        let tag = std::str::from_utf8(&tag)
+            .map_err(|_| invalid_data!("couldn't convert the tag to a utf8 string"))?;
+
+        Ok(match tag {
+            "??" => Self::UnknownMsg,
+            "AV" => Self::AgreementVote,
+            "MI" => Self::MsgOfInterest,
+            "MS" => Self::MsgDigestSkip,
+            "NP" => Self::NetPrioResponse,
+            "pi" => Self::Ping,
+            "pj" => Self::PingReply,
+            "PP" => Self::ProposalPayload,
+            "SP" => Self::StateProofSig,
+            "TS" => Self::TopicMsgResp,
+            "TX" => Self::Txn,
+            "UC" => Self::UniCatchupReq,
+            "UE" => Self::UniEnsBlockReq,
+            "VB" => Self::VoteBundle,
+            _ => return Err(invalid_data!("unexpected tag")),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Payload {
+    MsgOfInterest(MsgOfInterest),
+    NotImplemented,
+}
+
+/// [MsgOfInterest] contains a tag list in which the node is interested.
+#[derive(Debug)]
+pub struct MsgOfInterest {
+    /// Message tags for which the node is interested (subscribed).
+    // TODO(Rqnsom): convert to a hashset.
+    pub tags: String,
+}
+
+impl MsgOfInterest {
+    /// Decodes a byte stream from the [input] using the [topic] codec to generate a [MsgOfInterest] message.
+    fn from(input: &mut BytesMut, topic: &mut TopicCodec) -> Result<Self, io::Error> {
```

Comment:
we need to change this name so as not to confuse it with the typical `From<T>` convention

Reply:
How does the `construct` sound?

Reply:
couldn't this be part of `decode` for the corresponding codec? that would be the most idiomatic way

Reply:
Sooooo. This is the only thing I wasn't sure how to handle.

It's part of TagMsgCodec still. But handled in a separate `construct` (prev `from`) function.

Reply:
I'm still unsure about this; I feel like we should be able to do this from within the `TopicCodec`, where it would return `MsgOfInterest` instead of the current `Vec<Topic>` :thinking:

Reply:
Okay, I've done it like that

But `MsgOfInterest`'s definition is kept in the `payload.rs` (codec) file. But maybe it's better to move the definition into topic.rs 🤔
Because I know that the `MsgPack` message (payload) definitions will be in the `msgpack.rs`.